### PR TITLE
Use bindgen to generate bindings

### LINKED ISSUE #2: Use bindgen to generate -sys bindigs

### PR DESCRIPTION
**Pull request by panicbit**
Fixes #2

---
This change is [Reviewable](https://reviewable.io/reviews/hercules-team/rust-augeas/9)

### CODE REVIEW

**Comment by panicbit** [comment]
I'd like to wait until travis integration is enabled before merging this.

**Comment by lutter** [comment]
Just saw this after posting my PR. I guess a 'jinx' is in order. We'll have to figure out a way to reconcile the two PR's; I am not against using bindgen, but when I played with it it generated a ton of additional stuff because of the `aug_to_xml` method, which I don't think we need in the Rust bindings (plus, I'd like to avoid depending on libxml2)

We definitely need the changes you made to `AugFlag`

I'll look through your PR a little more tomorrow.

**Comment by panicbit** [comment]
We can blacklist `aug_to_xml` via `.blacklist('aug_to_xml')`, which would get rid of all the related xml junk. :smiley:

File: augeas_sys/build.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+extern crate bindgen;
+extern crate pkg_config;
+
+use std::env;
+use std::path::PathBuf;
+
+fn main() {
+    let augeas = pkg_config::Config::new()
+        .atleast_version("1.11.0")
```
Comment (lutter) [inline augeas_sys/build.rs]: Could we downgrade that to 1.10.1 ? For a bunch of dumb reasons, Fedora doesn't have 1.11, and there are no API changes between 1.10 and 1.11, so the downgraded crate will work with either.

File: src/lib.rs [hunk 2]
```diff
@@ -1,85 +1,84 @@
...
-        }
-    }
-
-    pub fn new(root: &str, loadpath: &str, flags: AugFlag) -> Result<Augeas> {
-        let root_c = try!(CString::new(root));
-        let loadpath_c = try!(CString::new(loadpath));
-
-        let augeas = unsafe {
-            raw::aug_init(root_c.as_ptr(), loadpath_c.as_ptr(), flags as u32)
+    pub fn init<'a>(root: impl Into<Option<&'a str>>, loadpath: &str, flags: Flags) -> Result<Self> {
```
Comment (lutter) [inline src/lib.rs]: `loadpath` should be optional, too,

File: src/lib.rs [hunk 3]
```diff
@@ -1,85 +1,84 @@
...
-    fn make_result<T>(&self, v : T) -> Result<T> {
-        let err = unsafe { raw::aug_error(self.ptr) };
-        if err == raw::ErrorCode::NoError {
-            Ok(v)
-        } else {
-            self.make_error()
-        }
-    }
-
-    pub fn new(root: &str, loadpath: &str, flags: AugFlag) -> Result<Augeas> {
```
Comment (lutter) [inline src/lib.rs]: Why change the name from `new` to `init` ? I thought `new` was the more common name for something like this in Rust.

**Comment by panicbit** [comment]
> One other thought, though that could be done later: would it be friendlier to users to check bindings.rs in so that people do not have to have Augeas headers (and headers for its deps) installed ?

I feel like it would be friendlier to automatically have the latest version of the augeas API available. There are also some possible cross-platform issues like [this one](https://github.com/rust-lang-nursery/rust-bindgen/issues/1213).

File: augeas_sys/build.rs (hunk 1 above)
Comment (panicbit) [reply augeas_sys/build.rs]: Yup

File: src/lib.rs (hunk 2 above)
Comment (panicbit) [reply src/lib.rs]: Hm, the doc doesn't mention explicitly that it is optional. Can it be NULL?
At any rate, I'll revisit these things for #5 anyway.

Comment (lutter) [reply src/lib.rs]: Yes, NULL is allowed there, and the doc doesn't mention that. Ugh. I'll fix up the docs in `augeas.h`.

File: src/lib.rs (hunk 3 above)
Comment (lutter) [reply src/lib.rs]: Sounds good.

**Comment by lutter** [comment]
I agree, there is definitely the potential for surprises there. So, better to leave it how you have it right now.

**Comment by panicbit** [comment]
To avoid duplicated work like this in the future I suggest we assign ourself to an issue when we start working on one.

**Comment by lutter** [comment]
Yes, absolutely. I'll rework my PR on top of yours and update it to expose the API calls that your bindgen work pulled in.

**Pull request merged**